File: sdk/flutter/ios/Classes/GoMailerPlugin.swift
import Flutter
import UIKit
import UserNotifications

public class GoMailerPlugin: NSObject, FlutterPlugin, UNUserNotificationCenterDelegate, FlutterStreamHandler {
  var deviceTokenResult: FlutterResult?
  var email: String?
  var apiKey: String?
  var currentDeviceToken: String?
  var baseUrl: String = "https://api.go-mailer.com/v1"
  private var eventSink: FlutterEventSink?
  // Reliability state
  private var pendingEvents: [(name: String, properties: [String: Any]?, email: String)] = []
  private let backoffMaxAttempts = 5
  private let backoffBaseDelay: Double = 0.5 // seconds
  private let backoffFactor: Double = 2.0
  private let backoffJitter: Double = 0.15 // seconds
  private let sdkVersion = "1.3.0"
  private let maxPendingEvents = 100

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "go_mailer", binaryMessenger: registrar.messenger())
    let instance = GoMailerPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
    
  // Register event channel for push notification events
  let eventChannel = FlutterEventChannel(name: "go_mailer_events", binaryMessenger: registrar.messenger())
  eventChannel.setStreamHandler(instance)
    
    NotificationCenter.default.addObserver(
      instance,
      selector: #selector(instance.didRegisterForRemoteNotifications(_:)),
      name: NSNotification.Name("GoMailerDidRegisterForRemoteNotifications"),
      object: nil
    )
    
    NotificationCenter.default.addObserver(
      instance,
      selector: #selector(instance.didFailToRegisterForRemoteNotifications(_:)),
      name: NSNotification.Name("GoMailerDidFailToRegisterForRemoteNotifications"),
      object: nil
    )
    // Restore any persisted queued events
    instance.restorePendingEventsIfNeeded()
  }

  // MARK: - FlutterStreamHandler
  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    debugPrint("📬 EventChannel: onListen called")
    self.eventSink = events
    emitEvent(type: "stream_ready", data: [:])
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    debugPrint("📬 EventChannel: onCancel called")
    self.eventSink = nil
    return nil
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
  switch call.method {
    case "initialize":
      debugPrint("🚀 Initializing Go Mailer plugin")
      if let args = call.arguments as? [String: Any] {
        self.apiKey = args["apiKey"] as? String
        if let key = self.apiKey {
          let masked = key.count > 6 ? String(key.prefix(6)) + "***" : key
          debugPrint("✅ API Key set (masked): \(masked)")
        } else {
          debugPrint("✅ API Key set (masked): null")
        }
        
        if let config = args["config"] as? [String: Any],
           let configBaseUrl = config["baseUrl"] as? String {
          self.baseUrl = configBaseUrl
          debugPrint("🌐 Base URL set: \(self.baseUrl)")
        }
      }
  emitEvent(type: "initialized", data: ["baseUrl": baseUrl, "email": email ?? ""])    
  result(nil)
      
    case "registerForPushNotifications":
      debugPrint("📱 Registering for push notifications")
      self.deviceTokenResult = result
      if let args = call.arguments as? [String: Any] {
        self.email = args["email"] as? String
  debugPrint("📧 Email set for registration: \(maskEmail(self.email))")
      }
      
      UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .badge, .sound]) { granted, error in
        if granted {
          debugPrint("✅ Notification permission granted")
          DispatchQueue.main.async {
            UIApplication.shared.registerForRemoteNotifications()
            let center = UNUserNotificationCenter.current()
            center.delegate = self
          }
        } else {
          debugPrint("❌ Notification permission denied: \(error?.localizedDescription ?? "Unknown")")
          result(FlutterError(
            code: "PERMISSION_DENIED",
            message: "User denied notification permission",
            details: error?.localizedDescription
          ))
        }
      }
      
    case "setUser":
      debugPrint("👤 Setting user data")
      if let args = call.arguments as? [String: Any] {
        self.email = args["email"] as? String
  debugPrint("📧 User email set: \(maskEmail(self.email))")
      }
      result(nil)
      
    case "getDeviceToken":
      debugPrint("🔑 Getting device token")
      if let token = currentDeviceToken {
        debugPrint("✅ Returning stored device token: \(token.prefix(20))...")
        result(token)
      } else {
        debugPrint("⚠️ No device token available yet")
        result(nil)
      }
      
    case "trackEvent":
      debugPrint("📊 Tracking event")
      if let args = call.arguments as? [String: Any] {
        let eventName = args["eventName"] as? String ?? "unknown"
        let properties = args["properties"] as? [String: Any]
  debugPrint("📊 Event: \(eventName), Properties: \(properties ?? [:]), email=\(maskEmail(self.email))")
        queueOrSendEvent(name: eventName, properties: properties, email: self.email ?? "")
      }
      result(nil)
      
    case "setUserAttributes":
      debugPrint("🏷️ Setting user attributes")
      result(nil)
      
    case "addUserTags":
      debugPrint("➕ Adding user tags")
      result(nil)
      
    case "removeUserTags":
      debugPrint("➖ Removing user tags")
      result(nil)
      
    case "setAnalyticsEnabled":
      debugPrint("📈 Setting analytics enabled")
      result(nil)
      
    case "setLogLevel":
      debugPrint("📝 Setting log level")
      result(nil)
      
    case "flushPendingEvents":
      flushPendingEvents()
      result(nil)
    case "getSdkInfo":
      let info: [String: Any] = [
        "version": sdkVersion,
        "baseUrl": baseUrl,
        "email": email ?? "",
        "deviceToken": currentDeviceToken ?? ""
      ]
      result(info)
    default:
      debugPrint("❌ Method not implemented: \(call.method)")
      result(FlutterMethodNotImplemented)
    }
  }

  @objc func didRegisterForRemoteNotifications(_ notification: Notification) {
    guard let deviceToken = notification.userInfo?["deviceToken"] as? Data else {
      debugPrint("❌ No device token in notification")
      deviceTokenResult?(FlutterError(code: "NO_TOKEN", message: "No device token", details: nil))
      return
    }
    
    let tokenParts = deviceToken.map { data in String(format: "%02.2hhx", data) }
    let token = tokenParts.joined()
    
    currentDeviceToken = token
    debugPrint("✅ Device token received: \(token.prefix(20))...")
    
    deviceTokenResult?(token)
    deviceTokenResult = nil
    
    // Send token event to Flutter if listener is active
    emitEvent(type: "registered", data: ["deviceToken": token, "email": email ?? ""])    
    
    sendDeviceTokenToServer(token: token, email: self.email ?? "")
  }

  @objc func didFailToRegisterForRemoteNotifications(_ notification: Notification) {
    let error = notification.userInfo?["error"] as? Error
    deviceTokenResult?(
      FlutterError(
        code: "REGISTER_FAILED",
        message: "Failed to register for remote notifications",
        details: error?.localizedDescription
      )
    )
    deviceTokenResult = nil
    
    emitEvent(type: "register_failed", data: ["error": error?.localizedDescription ?? "unknown"])    
  }

  func sendDeviceTokenToServer(token: String, email: String) {
    let bundleId = Bundle.main.bundleIdentifier ?? "unknown.bundle"
    let payload: [String: Any] = [
      "email": email,
      "maskedEmail": maskEmail(email),
      "gm_mobi_push": [
        "deviceToken": token,
        "platform": "ios",
        "bundleId": bundleId,
        "appVersion": Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "Unknown",
        "timestamp": ISO8601DateFormatter().string(from: Date())
      ]
    ]
    let url = URL(string: "\(baseUrl)/contacts")!
    var req = URLRequest(url: url)
    req.httpMethod = "POST"
    req.setValue("application/json", forHTTPHeaderField: "Content-Type")
    if let apiKey = apiKey { req.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization") }
    req.httpBody = try? JSONSerialization.data(withJSONObject: payload)
    performRequestWithBackoff(label: "token_registration", requestBuilder: { return req }) { success, terminal in
      if !success && terminal {
        self.emitEvent(type: "token_failed", data: [:])
      } else if success {
        // Flush any queued events now that token registered
        self.flushPendingEvents()
      }
    }
  }
  
  func sendEventToServer(eventName: String, properties: [String: Any]?, email: String) {
    let bundleId = Bundle.main.bundleIdentifier ?? "unknown.bundle"
    let body: [String: Any] = [
      "email": email,
      "maskedEmail": maskEmail(email),
      "eventName": eventName,
      "properties": properties ?? [:],
      "platform": "ios",
      "bundleId": bundleId,
      "appVersion": Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "Unknown",
      "timestamp": ISO8601DateFormatter().string(from: Date())
    ]
    let url = URL(string: "\(baseUrl)/events/push")!
    var req = URLRequest(url: url)
    req.httpMethod = "POST"
    req.setValue("application/json", forHTTPHeaderField: "Content-Type")
    if let apiKey = apiKey { req.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization") }
    req.httpBody = try? JSONSerialization.data(withJSONObject: body)
    performRequestWithBackoff(label: "event_\(eventName)", requestBuilder: { return req }) { success, terminal in
      if success {
        self.emitEvent(type: "event_tracked", data: ["event": eventName])
      } else if terminal {
        self.emitEvent(type: "event_failed", data: ["event": eventName])
      }
    }
  }
  
  // MARK: - UNUserNotificationCenterDelegate
  
  public func userNotificationCenter(
    _ center: UNUserNotificationCenter,
    willPresent notification: UNNotification,
    withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void
  ) {
    debugPrint("📱 Notification received while app is in foreground")
    completionHandler([.alert, .badge, .sound])
  }
  
  public func userNotificationCenter(
    _ center: UNUserNotificationCenter,
    didReceive response: UNNotificationResponse,
    withCompletionHandler completionHandler: @escaping () -> Void
  ) {
    debugPrint("📱 Notification tapped")
    
  let notification = response.notification
    let userInfo = notification.request.content.userInfo
    let notificationId = userInfo["notification_id"] as? String ?? "unknown"
    let title = notification.request.content.title
    let body = notification.request.content.body
    
    debugPrint("📱 Notification details - ID: \(notificationId), Title: \(title ?? "Unknown")")
    
    let properties: [String: Any] = [
      "notification_id": notificationId,
      "notification_title": title ?? "Unknown",
      "notification_body": body ?? "",
      "action_identifier": response.actionIdentifier,
      "user_info": userInfo,
      "click_timestamp": ISO8601DateFormatter().string(from: Date())
    ]
    
    sendEventToServer(eventName: "notification_clicked", properties: properties, email: self.email ?? "")

    // Forward event to Flutter listeners
    emitEvent(type: "notification_clicked", data: [
      "id": notificationId,
      "title": title ?? "",
      "body": body ?? "",
      "userInfo": userInfo
    ])
    
    completionHandler()
  }
}

// MARK: - Reliability Helpers
extension GoMailerPlugin {
  private func emitEvent(type: String, data: [String: Any]) {
    guard let sink = eventSink else { return }
    DispatchQueue.main.async { sink(["type": type, "data": data]) }
  }

  private func queueOrSendEvent(name: String, properties: [String: Any]?, email: String) {
    if currentDeviceToken == nil { // queue
      var dropped: (name: String, properties: [String: Any]?, email: String)? = nil
      if pendingEvents.count >= maxPendingEvents {
        dropped = pendingEvents.removeFirst()
      }
      pendingEvents.append((name: name, properties: properties, email: email))
      persistPendingEvents()
      if let d = dropped { emitEvent(type: "event_dropped", data: ["event": d.name, "reason": "queue_full"]) }
      emitEvent(type: "event_queued", data: ["event": name])
    } else {
      sendEventToServer(eventName: name, properties: properties, email: email)
    }
  }

  private func maskEmail(_ email: String?) -> String {
    guard let email = email, !email.isEmpty else { return "" }
    let parts = email.split(separator: "@")
    guard let first = parts.first else { return "" }
    let local = String(first)
    let maskedLocal: String
    switch local.count {
      case 0: maskedLocal = ""
      case 1: maskedLocal = "*"
      case 2: maskedLocal = String(local.prefix(1)) + "*"
      default: maskedLocal = String(local.prefix(2)) + "***"
    }
    if parts.count > 1 { return maskedLocal + "@" + parts[1] } else { return maskedLocal }
  }

  func flushPendingEvents() {
    guard currentDeviceToken != nil, !pendingEvents.isEmpty else { return }
    let toSend = pendingEvents; pendingEvents.removeAll(); persistPendingEvents()
    toSend.forEach { tuple in
      sendEventToServer(eventName: tuple.name, properties: tuple.properties, email: tuple.email)
    }
  }

  private func performRequestWithBackoff(label: String, requestBuilder: @escaping () -> URLRequest, completion: @escaping (_ success: Bool, _ terminal: Bool) -> Void) {
    func attempt(_ n: Int) {
      let req = requestBuilder() // copy to avoid capture semantics issues
      let currentAttempt = n
      let task = URLSession.shared.dataTask(with: req) { data, response, error in
        if let error = error {
          if currentAttempt >= self.backoffMaxAttempts { completion(false, true); return }
          scheduleRetry(n: currentAttempt)
          return
        }
        let status = (response as? HTTPURLResponse)?.statusCode ?? 0
        if (200..<300).contains(status) {
          completion(true, false)
        } else {
          // retry on 429 or 5xx
          if (status == 429 || status >= 500) && currentAttempt < self.backoffMaxAttempts {
            scheduleRetry(n: currentAttempt)
          } else {
            completion(false, true)
          }
        }
      }
      task.resume()
    }
    func scheduleRetry(n: Int) {
      let base = self.backoffBaseDelay * pow(self.backoffFactor, Double(n - 1))
      let jitter = Double.random(in: 0...self.backoffJitter)
      let delay = min(base + jitter, 10.0)
      DispatchQueue.global().asyncAfter(deadline: .now() + delay) {
        attempt(n + 1)
      }
    }
    attempt(1)
  }

  // MARK: - Persistence (UserDefaults)
  private var pendingKey: String { return "gomailer.pending.events" }

  private func persistPendingEvents() {
    do {
      let arr: [[String: Any]] = pendingEvents.map { e in
        [
          "name": e.name,
          "email": e.email,
          "properties": e.properties ?? [:]
        ]
      }
      let data = try JSONSerialization.data(withJSONObject: arr, options: [])
      UserDefaults.standard.set(data, forKey: pendingKey)
    } catch {
      debugPrint("⚠️ Failed to persist pending events: \(error)")
    }
  }

  private func restorePendingEventsIfNeeded() {
    guard pendingEvents.isEmpty, let data = UserDefaults.standard.data(forKey: pendingKey) else { return }
    do {
      let raw = try JSONSerialization.jsonObject(with: data, options: []) as? [[String: Any]] ?? []
      let restored = raw.map { dict -> (name: String, properties: [String: Any]?, email: String) in
        let name = dict["name"] as? String ?? "unknown"
        let email = dict["email"] as? String ?? ""
        let props = dict["properties"] as? [String: Any]
        return (name: name, properties: props, email: email)
      }
      if !restored.isEmpty {
        pendingEvents.append(contentsOf: restored)
        debugPrint("🔁 Restored \(restored.count) pending events from disk")
      }
    } catch {
      debugPrint("⚠️ Failed to restore pending events: \(error)")
    }
  }
}

// Ensure restoration on class load
extension GoMailerPlugin {
  static func initializePersistence() { /* reserved */ }
}

File: sdk/flutter/android/src/main/kotlin/com/gomailer/go_mailer/GoMailerPlugin.kt
package com.gomailer.go_mailer

import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodChannel.Result
import android.util.Log
import android.content.Context
import android.content.SharedPreferences
import kotlinx.coroutines.*
import kotlinx.coroutines.tasks.await
import org.json.JSONObject
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.concurrent.TimeUnit
import java.text.SimpleDateFormat
import java.util.*

/** GoMailerPlugin */
class GoMailerPlugin: FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel : MethodChannel
  private lateinit var eventChannel: EventChannel
  private var apiKey: String? = null
  private var baseUrl: String = "https://api.go-mailer.com/v1"
  private var currentEmail: String? = null
  private var deviceToken: String? = null
  private var eventsSink: EventChannel.EventSink? = null
  // Queue for events attempted before token registration or during transient failures
  private val pendingEvents = Collections.synchronizedList(mutableListOf<Triple<String, Map<*, *>?, String>>())
  private val MAX_PENDING_EVENTS = 100

  private val jsonMedia = "application/json; charset=utf-8".toMediaType()
  private val client = OkHttpClient.Builder()
    .connectTimeout(10, TimeUnit.SECONDS)
    .readTimeout(15, TimeUnit.SECONDS)
    .writeTimeout(15, TimeUnit.SECONDS)
    .retryOnConnectionFailure(true)
    .build()

  private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
  private var appContext: android.content.Context? = null
  private var prefs: SharedPreferences? = null
  private val PREFS_NAME = "gomailer_sdk_prefs"
  private val KEY_PENDING_EVENTS = "pending_events_json"

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "go_mailer")
    channel.setMethodCallHandler(this)
    eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "go_mailer_events")
    eventChannel.setStreamHandler(this)
  appContext = flutterPluginBinding.applicationContext
  prefs = appContext?.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
  // Load any persisted pending events
  restorePendingEvents()
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    when (call.method) {
      "initialize" -> {
        Log.d("GoMailer", "🚀 Initializing Go Mailer plugin")
        val args = call.arguments as? Map<*, *>
  apiKey = args?.get("apiKey") as? String
  val masked = apiKey?.let { if (it.length > 6) it.substring(0,6) + "***" else it } ?: "null"
  Log.d("GoMailer", "✅ API Key set (masked): $masked")
        
        // Extract baseUrl from config if provided
        val config = args?.get("config") as? Map<*, *>
        val configBaseUrl = config?.get("baseUrl") as? String
        if (configBaseUrl != null) {
          baseUrl = configBaseUrl
          Log.d("GoMailer", "🌐 Base URL set: $baseUrl")
        }
        
        // Configure Firebase Messaging Service
        GoMailerFirebaseMessagingService.setConfig(apiKey, baseUrl, currentEmail)
        emitEvent("initialized", mapOf(
          "baseUrl" to baseUrl,
          "email" to (currentEmail ?: ""))
        )
        result.success(null)
      }
      "registerForPushNotifications" -> {
        Log.d("GoMailer", "📱 Registering for push notifications")
        val args = call.arguments as? Map<*, *>
  currentEmail = args?.get("email") as? String
  Log.d("GoMailer", "📧 Email set for registration: ${maskEmail(currentEmail)}")
        
        // Get FCM token and send to backend
        scope.launch {
          try {
            val token = com.google.firebase.messaging.FirebaseMessaging.getInstance().token.await()
            Log.d("GoMailer", "✅ FCM token obtained: ${token.take(20)}...")
            
            // Send token to backend
            sendTokenToServer(currentEmail ?: "", token)
            deviceToken = token
            emitEvent("registered", mapOf("deviceToken" to token, "email" to (currentEmail ?: "")))
            
            withContext(Dispatchers.Main) {
              result.success(token)
            }
          } catch (e: Exception) {
            Log.e("GoMailer", "❌ Failed to register for push notifications: ${e.message}")
            emitEvent("register_failed", mapOf("error" to (e.message ?: "unknown")))
            withContext(Dispatchers.Main) {
              result.success(null)
            }
          }
        }
      }
      "setUser" -> {
        Log.d("GoMailer", "👤 Setting user data")
        val args = call.arguments as? Map<*, *>
  currentEmail = args?.get("email") as? String
  Log.d("GoMailer", "📧 User email set: ${maskEmail(currentEmail)}")
        
        // Update Firebase Messaging Service with new email
        GoMailerFirebaseMessagingService.setConfig(apiKey, baseUrl, currentEmail)
        
        result.success(null)
      }
      "trackEvent" -> {
        Log.d("GoMailer", "📊 Tracking event")
        val args = call.arguments as? Map<*, *>
        val eventName = args?.get("eventName") as? String ?: "unknown"
        val properties = args?.get("properties") as? Map<*, *>
  Log.d("GoMailer", "📊 Event: $eventName, Properties: $properties, email=${maskEmail(currentEmail)}")
        queueOrSendEvent(eventName, properties, currentEmail ?: "")
        result.success(null)
      }
      "trackNotificationClick" -> {
        Log.d("GoMailer", "👆 Tracking notification click")
        val args = call.arguments as? Map<*, *>
        val notificationId = args?.get("notificationId") as? String ?: "unknown"
        val title = args?.get("title") as? String ?: "Unknown"
        val body = args?.get("body") as? String ?: ""
        val email = args?.get("email") as? String ?: currentEmail ?: ""
        
        val properties = mapOf(
          "notification_id" to notificationId,
          "notification_title" to title,
          "notification_body" to body,
          "clicked_timestamp" to SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US).format(Date()),
          "platform" to "android"
        )
        
        Log.d("GoMailer", "👆 Notification click: $notificationId (queueOrSend)")
        queueOrSendEvent("notification_clicked", properties, email)
        result.success(null)
      }
      "getDeviceToken" -> {
        Log.d("GoMailer", "🔑 Getting device token")
        scope.launch {
          try {
            val token = com.google.firebase.messaging.FirebaseMessaging.getInstance().token.await()
            Log.d("GoMailer", "✅ FCM token retrieved: ${token.take(20)}...")
            withContext(Dispatchers.Main) {
              result.success(token)
            }
          } catch (e: Exception) {
            Log.e("GoMailer", "❌ Failed to get FCM token: ${e.message}")
            withContext(Dispatchers.Main) {
              result.success(null)
            }
          }
        }
      }
      "getSdkInfo" -> {
        val info = mapOf(
          "version" to "1.3.0",
          "baseUrl" to baseUrl,
          "email" to (currentEmail ?: ""),
          "deviceToken" to (deviceToken ?: "")
        )
        result.success(info)
      }
      "flushPendingEvents" -> {
        scope.launch { flushPendingEvents() }
        result.success(null)
      }
      else -> result.notImplemented()
    }
  }

  private fun sendTokenToServer(email: String, token: String) {
    scope.launch { performWithBackoff("token_registration") { attempt ->
      try {
  val bundleId = appContext?.packageName ?: "com.gomailer.go_mailer"
        val json = JSONObject().apply {
          put("email", email)
          put("maskedEmail", maskEmail(email))
          put("gm_mobi_push", JSONObject().apply {
            put("deviceToken", token)
            put("platform", "android")
            put("bundleId", bundleId)
            put("appVersion", "1.3.0")
            put("timestamp", isoUtc())
          })
        }
        val req = Request.Builder()
          .url("$baseUrl/contacts")
          .header("Content-Type", "application/json")
          .apply { apiKey?.let { header("Authorization", "Bearer $it") } }
          .post(json.toString().toRequestBody(jsonMedia))
          .build()
        client.newCall(req).execute().use { resp ->
          val code = resp.code
          val bodyStr = resp.body?.string()
          Log.d("GoMailer", "📤 Device token POST attempt=$attempt status=$code")
          return@performWithBackoff if (resp.isSuccessful) {
            true // Completed successfully, do not retry
          } else {
            val retryable = code == 429 || code >= 500
            if (!retryable) {
              emitEvent("token_failed", mapOf("status" to code, "body" to (bodyStr ?: "")))
            }
            retryable // true means should retry, false means stop
          }
        }
      } catch (e: Exception) {
        Log.e("GoMailer", "❌ Token POST error attempt=$attempt : ${e.message}")
        return@performWithBackoff true // network or transient error -> retry
      }
    } ?: run {
      emitEvent("token_failed", mapOf("error" to "exhausted_retries"))
    } }
  }

  private fun sendEventToServer(eventName: String, properties: Map<*, *>?, email: String) {
    scope.launch { performWithBackoff("event_$eventName") { attempt ->
      try {
  val bundleId = appContext?.packageName ?: "com.gomailer.go_mailer"
        val json = JSONObject().apply {
          put("email", email)
          put("maskedEmail", maskEmail(email))
          put("eventName", eventName)
          put("properties", properties ?: JSONObject())
          put("platform", "android")
          put("bundleId", bundleId)
          put("appVersion", "1.3.0")
          put("timestamp", isoUtc())
        }
        val req = Request.Builder()
          .url("$baseUrl/events/push")
          .header("Content-Type", "application/json")
          .header("Authorization", "Bearer $apiKey")
          .post(json.toString().toRequestBody(jsonMedia))
          .build()
        client.newCall(req).execute().use { resp ->
          val code = resp.code
          Log.d("GoMailer", "📡 Event POST '$eventName' attempt=$attempt status=$code")
          return@performWithBackoff if (resp.isSuccessful) {
            emitEvent("event_tracked", mapOf("event" to eventName))
            true
          } else {
            val retryable = code == 429 || code >= 500
            if (!retryable) {
              emitEvent("event_failed", mapOf("event" to eventName, "status" to code))
            }
            retryable
          }
        }
      } catch (e: Exception) {
        Log.e("GoMailer", "❌ Event POST error '$eventName' attempt=$attempt : ${e.message}")
        return@performWithBackoff true
      }
    } ?: run {
      emitEvent("event_failed", mapOf("event" to eventName, "error" to "exhausted_retries"))
    } }
  }

  private suspend fun performWithBackoff(
    label: String,
    maxRetries: Int = 5,
    initialDelayMs: Long = 500,
    factor: Double = 2.0,
    jitterMs: Long = 150,
    block: suspend (attempt: Int) -> Boolean // true means retry, false means completed (success or terminal)
  ): Boolean? {
    var delayMs = initialDelayMs
    for (attempt in 1..maxRetries) {
      val shouldRetry = block(attempt)
      if (!shouldRetry) return true
      if (attempt < maxRetries) {
        val jitter = (0..jitterMs).random()
        val total = delayMs + jitter
        Log.d("GoMailer", "⏳ Backoff label=$label attempt=$attempt waiting=${total}ms")
        delay(total)
        delayMs = (delayMs * factor).toLong().coerceAtMost(10_000)
      }
    }
    return null
  }

  private fun queueOrSendEvent(eventName: String, properties: Map<*, *>?, email: String) {
    if (deviceToken == null) {
      var dropped: Triple<String, Map<*, *>?, String>? = null
      synchronized(pendingEvents) {
        if (pendingEvents.size >= MAX_PENDING_EVENTS) {
          dropped = pendingEvents.removeAt(0)
        }
        pendingEvents.add(Triple(eventName, properties, email))
      }
      if (dropped != null) {
        Log.w("GoMailer", "🗑️ Dropping oldest queued event '${dropped!!.first}' due to capacity $MAX_PENDING_EVENTS")
        emitEvent("event_dropped", mapOf("event" to dropped!!.first, "reason" to "queue_full"))
      }
      Log.d("GoMailer", "🕓 Queued event '$eventName' (device token pending)")
      emitEvent("event_queued", mapOf("event" to eventName))
      persistPendingEvents()
    } else {
      sendEventToServer(eventName, properties, email)
    }
  }

  private suspend fun flushPendingEvents() {
    if (deviceToken == null || pendingEvents.isEmpty()) return
    val toFlush = synchronized(pendingEvents) { val copy = pendingEvents.toList(); pendingEvents.clear(); copy }
    Log.d("GoMailer", "🚚 Flushing ${toFlush.size} queued events")
    toFlush.forEach { (name, props, email) -> sendEventToServer(name, props, email) }
    persistPendingEvents() // clear persisted copy
  }

  private fun isoUtc(): String = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US).apply {
    timeZone = TimeZone.getTimeZone("UTC")
  }.format(Date())

  private fun emitEvent(type: String, data: Map<String, Any?>) {
    eventsSink?.success(mapOf("type" to type, "data" to data))
  }

  private fun maskEmail(email: String?): String {
    if (email.isNullOrEmpty()) return ""
    val parts = email.split("@")
    val local = parts[0]
    val maskedLocal = when {
      local.length <= 1 -> "*"
      local.length == 2 -> local.first() + "*"
      else -> local.substring(0, 2) + "***"
    }
    return if (parts.size > 1) maskedLocal + "@" + parts[1] else maskedLocal
  }

  // EventChannel.StreamHandler
  override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
    eventsSink = events
    emitEvent("stream_ready", emptyMap())
  }

  override fun onCancel(arguments: Any?) {
    eventsSink = null
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
    eventChannel.setStreamHandler(null)
    scope.cancel()
  }

  // Persistence helpers
  private fun persistPendingEvents() {
    try {
      val arr = org.json.JSONArray()
      synchronized(pendingEvents) {
        pendingEvents.forEach { (name, props, email) ->
          val obj = org.json.JSONObject()
          obj.put("name", name)
          obj.put("email", email)
          obj.put("properties", if (props != null) org.json.JSONObject(props) else org.json.JSONObject())
          arr.put(obj)
        }
      }
      prefs?.edit()?.putString(KEY_PENDING_EVENTS, arr.toString())?.apply()
    } catch (e: Exception) {
      Log.w("GoMailer", "⚠️ Failed to persist pending events: ${e.message}")
    }
  }

  private fun restorePendingEvents() {
    try {
      val raw = prefs?.getString(KEY_PENDING_EVENTS, null) ?: return
      val arr = org.json.JSONArray(raw)
      for (i in 0 until arr.length()) {
        val obj = arr.getJSONObject(i)
        val name = obj.getString("name")
        val email = obj.getString("email")
        val propsObj = obj.optJSONObject("properties")
        val map: Map<String, Any?>? = propsObj?.let { json ->
          val keys = json.keys()
          val temp = mutableMapOf<String, Any?>()
          while (keys.hasNext()) {
            val k = keys.next()
            temp[k] = json.get(k)
          }
            temp
        }
        pendingEvents.add(Triple(name, map, email))
      }
      if (pendingEvents.isNotEmpty()) {
        Log.d("GoMailer", "🔁 Restored ${pendingEvents.size} pending events from disk")
      }
    } catch (e: Exception) {
      Log.w("GoMailer", "⚠️ Failed to restore pending events: ${e.message}")
    }
  }
}
